File: neetcode150/medium/cpp/143-Reorder-List.cpp
#include <algorithm>

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        // Reverse fast
        ListNode* curr = slow->next;
        ListNode* nxt = NULL;
        ListNode* prev = NULL;
        while (curr != NULL) {
            nxt = curr->next;
            curr->next = prev;
            prev = curr;
            curr = nxt;
        }
        slow->next = NULL;

        // Merge
        ListNode* first = head; 
        ListNode* second = prev; 

        while (second != NULL) {
            ListNode* first_nxt = first->next;
            ListNode* second_nxt = second->next;

            first->next = second;
            second->next = first_nxt;
            first = first_nxt;
            second = second_nxt;
        }
    }
};

File: neetcode150/medium/143-Reorder-List.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        # reverse fast
        nxt, prev = None, None
        curr = slow.next
        slow.next = None
        while curr != None:
            nxt = curr.next
            curr.next = prev 
            prev = curr
            curr = nxt
        
        # merge head and prev
        first, second = head, prev

        while second:
            first_nxt = first.next
            second_nxt = second.next

            first.next = second
            second.next = first_nxt
            first = first_nxt
            second = second_nxt